tests_e2e/playwright/navigator.py

```diff
@@ -0,0 +1,448 @@
+            # select instance type (this will fail if this instance type is not
+            # available)
+            self.page.locator(f"#profile-item-{self.instance_name}").click()
+            self.page.get_by_role("button", name="Start").click()
+
+            # # wait for server to spinup
+            # server_message = self.page.get_by_text(
+            #     f"Server ready at /user/{self.username}/",
+            #     exact=True,
+            # )
+            # server_message.wait_for(
+            #     timeout=self.wait_for_server_spinup,
+            #     state="attached",
+            # )
+
+        except Exception:
+            # if the server is already running
+            start_locator = self.page.get_by_role(
+                "button",
+                name="My Server",
+                exact=True,
+            )
+            start_locator.wait_for(timeout=3000, state="attached")
+            start_locator.click()
+
+        # wait for server spinup
+        self.page.wait_for_url(
+            f"{self.nebari_url}user/{self.username}/*",
+            wait_until="networkidle",
+        )
+
+        # the jupyter page loads independent of network activity so here
+        # we wait for the File menu to be available on the page, a proxy for
+        # the jupyterlab page being loaded.
+        file_locator = self.page.get_by_text("File", exact=True)
+        file_locator.wait_for(
+            timeout=self.wait_for_server_spinup,
+            state="attached",
+        )
+
+        logger.debug(">>> Sign in complete.")
+
+    def _check_for_kernel_popup(self):
+        """Is the kernel popup currently open?
+
+        Returns
+        -------
+        True if the kernel popup is open.
+        """
+        self.page.wait_for_load_state("networkidle")
+        visible = self.page.get_by_text("Select Kernel", exact=True).is_visible()
+
+        return visible
+
+    def reset_workspace(self):
+        """Reset the Jupyterlab workspace.
+
+        * Closes all Tabs & handle possible popups for saving changes,
+        * make sure any kernel popups are dealt with
+        * reset file browser is reset to root
+        * Finally, ensure that the Launcher screen is showing
+        """
+        logger.debug(">>> Reset JupyterLab workspace")
+
+        # server is already running and there is no popup
+        popup = self._check_for_kernel_popup()
+
+        # server is on running and there is a popup
+        if popup:
+            self._set_environment_via_popup(kernel=None)
+
+        # go to Kernel menu
+        kernel_menuitem = self.page.get_by_text("Kernel", exact=True)
+        kernel_menuitem.click()
+        # shut down multiple running kernels
+        try:
+            shut_down_all = self.page.get_by_text(
+                "Shut Down All Kernels...", exact=True
+            )
+            shut_down_all.wait_for(timeout=300, state="attached")
+            shut_down_all.click()
+        except Exception:
+            pass
```

Comment:
Nit:

```suggestion
        with contextlib.suppress(Exception):
            shut_down_all = self.page.get_by_text(
                "Shut Down All Kernels...", exact=True
            )
            shut_down_all.wait_for(timeout=300, state="attached")
            shut_down_all.click()
```

Saw this pattern multiple times. Suggestion applies everywhere.